File: backend/app/models/saved_search.py
"""SavedSearch SQLModel for BWARM Dashboard.

Represents user-defined search filters that can be saved and reused.
"""

from datetime import datetime
from typing import Any, Dict, Optional

from sqlalchemy import JSON, Column
from sqlmodel import Field, SQLModel


class SavedSearch(SQLModel, table=True):
    """Saved search entity for persisting filter configurations.

    Stores user-defined search criteria with flexible filter logic supporting
    AND/OR operations, field filters, and custom search parameters.
    """

    __tablename__ = "saved_searches"

    id: Optional[int] = Field(default=None, primary_key=True)

    # Ownership
    user_id: int = Field(foreign_key="users.id", index=True, nullable=False)

    # Search metadata
    name: str = Field(max_length=255, nullable=False)
    description: Optional[str] = Field(default=None, max_length=1000)

    # Filter configuration (stored as JSON)
    # Example structure:
    # {
    #   "logic": "AND",  # or "OR"
    #   "filters": [
    #     {"field": "has_iswc", "operator": "equals", "value": true},
    #     {"field": "title", "operator": "contains", "value": "love"}
    #   ],
    #   "sort": {"field": "created_at", "order": "desc"}
    # }
    filter_config: Dict[str, Any] = Field(
        default={},
        sa_column=Column(JSON, nullable=False),
    )

    # Preset indicator
    is_preset: bool = Field(default=False, index=True)

    # Usage tracking
    last_used_at: Optional[datetime] = Field(default=None)
    use_count: int = Field(default=0)

    # Favorites
    is_favorite: bool = Field(default=False, index=True)

    # Timestamps
    created_at: datetime = Field(
        default_factory=datetime.utcnow,
        nullable=False,
    )
    updated_at: datetime = Field(
        default_factory=datetime.utcnow,
        nullable=False,
        sa_column_kwargs={"onupdate": datetime.utcnow},
    )

    def __repr__(self) -> str:
        """String representation of SavedSearch."""
        return f"<SavedSearch(id={self.id}, name='{self.name}', user_id={self.user_id})>"

    class Config:
        """Pydantic model configuration."""

        json_schema_extra = {
            "example": {
                "name": "Works without ISWC",
                "description": "Find all works missing ISWC codes",
                "filter_config": {
                    "logic": "AND",
                    "filters": [
                        {"field": "has_iswc", "operator": "equals", "value": False},
                    ],
                },
                "is_favorite": False,
            }
        }

File: frontend/src/types/savedSearch.ts
/**
 * Saved Search Type Definitions
 */

export type FilterOperator =
  | "equals"
  | "not_equals"
  | "contains"
  | "not_contains"
  | "starts_with"
  | "ends_with"
  | "greater_than"
  | "less_than"
  | "greater_than_or_equal"
  | "less_than_or_equal"
  | "in"
  | "not_in"
  | "is_null"
  | "is_not_null";

export interface FilterCondition {
  field: string;
  operator: FilterOperator;
  value: any;
}

export interface FilterConfig {
  logic: "AND" | "OR";
  filters: FilterCondition[];
  sort?: {
    field: string;
    order: "asc" | "desc";
  };
}

export interface SavedSearch {
  id: number;
  user_id: number;
  name: string;
  description?: string;
  filter_config: FilterConfig;
  is_preset: boolean;
  is_favorite: boolean;
  last_used_at?: string;
  use_count: number;
  created_at: string;
  updated_at: string;
}

export interface SavedSearchCreate {
  name: string;
  description?: string;
  filter_config: FilterConfig;
  is_favorite?: boolean;
}

export interface SavedSearchUpdate {
  name?: string;
  description?: string;
  filter_config?: FilterConfig;
  is_favorite?: boolean;
}

export interface SavedSearchListResponse {
  total: number;
  searches: SavedSearch[];
}
